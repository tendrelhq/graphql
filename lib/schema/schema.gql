enum AuthenticationProvider {
  legacy
  clerk
}

enum LanguageCode {
  en
  es
}

type Language {
  id: ID!
  code: LanguageCode!
  name: Name!
  name_id: ID!
}

type Name {
  id: ID!
  language: Language!
  language_id: ID!
  value: String!
}

type Tag {
  id: ID!
  name: Name!
  name_id: ID!
  type: String!
  parent: Tag
  parent_id: ID
}

type Customer {
  id: ID!
  active: Boolean!
  name: Name!
  name_id: ID!
  defaultLanguage: Language!
  default_language_id: ID!
}

type Location {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  name: Name!
  name_id: ID!
  parent: Location
  parent_id: ID
  scan_code: ID
  site: Location!
  site_id: ID!
  tags: [Tag!]!

  children(options: LocationsQueryOptions): [Location!]!
}

type User {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  authentication_provider: Tag
  authentication_provider_id: ID
  language: Language!
  language_id: ID!
  name: String!
  tags: [Tag!]!
}

type Worker {
  id: ID!
  active: Boolean!
  activated_at: String
  deactivated_at: String
  language: Language!
  language_id: ID!
  role: Tag!
  role_id: ID!
  scan_code: String
  tags: [Tag!]!
  user: User!
  user_id: ID!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  languages: [Language!]!
  locations(customerId: ID!, options: LocationsQueryOptions): [Location!]!
  location(id: ID!): Location!
  roles: [Tag!]!
  workers(customerId: ID!, options: WorkersQueryOptions): [Worker!]!
}

input LocationsQueryOptions {
  cornerstone: Boolean
  site: Boolean
}

input WorkersQueryOptions {
  active: Boolean
  site: ID
}

type Mutation {
  # Location CRUD
  createLocation(input: CreateLocationInput!): Location!
  updateLocation(input: UpdateLocationInput!): Location!

  # User
  ## CRUD
  createUser(input: CreateUserInput!): User!
  updateUser(input: UpdateUserInput!): User!
  ## Helpers
  acceptInvite(input: AcceptInviteInput!): User!

  # Worker CRUD
  createWorker(input: CreateWorkerInput!): Worker!
  updateWorker(input: UpdateWorkerInput!): Worker!
}

input CreateLocationInput {
  id: ID
  active: Boolean!
  name: CreateNameInput!
  org_id: ID!
  scan_code: ID
}

input UpdateLocationInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  name: UpdateNameInput
  scan_code: ID
}

input CreateUserInput {
  id: ID
  active: Boolean!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input UpdateUserInput {
  id: ID!
  authentication_provider_id: ID
  authentication_identity: ID
  language_id: ID!
  name: String!
}

input CreateWorkerInput {
  id: ID
  active: Boolean!
  language_id: ID!
  org_id: ID!
  role_id: ID!
  scan_code: ID
  user_id: ID!
}

input UpdateWorkerInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  role_id: ID!
  scan_code: ID
}

input CreateNameInput {
  id: ID
  active: Boolean!
  language_id: ID!
  value: String!
}

input UpdateNameInput {
  id: ID!
  activated_at: String
  deactivated_at: String
  language_id: ID!
  value: String!
}

input AcceptInviteInput {
  id: ID!
  authentication_identity: ID!
}