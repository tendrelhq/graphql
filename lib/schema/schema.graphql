enum Language {
  en
  es
}

type Tag {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

type Customer {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

type Location {
  id: ID!
  customerId: ID!
  name: String!
  parentId: ID
  tags: [Tag!]!
}

type Query {
  customers: [Customer!]!
  customer(id: ID!): Customer!
  locations(customerId: ID!, parentId: ID): [Location!]!
  sites(customerId: ID!): [Location!]!
}

type Mutation {
  createCustomer(input: CreateCustomerInput!): ID!
  updateCustomer(input: UpdateCustomerInput!): ID!
  deleteCustomer(id: ID!): ID!
  createLocation(input: CreateLocationInput!): ID!
  updateLocation(input: UpdateLocationInput!): ID!
  deleteLocation(id: ID!): ID!
}

input CreateTagInput {
  id: ID!
  name: String!
  type: String!
  parent: ID
}

input UpdateTagInput {
  id: ID!
  name: String
}

input CreateCustomerInput {
  id: ID!
  name: String!
  defaultLanguage: Language!
}

input UpdateCustomerInput {
  id: ID!
  name: String
  defaultLanguage: Language
}

input CreateLocationInput {
  id: ID!
  customerId: ID!
  name: String!
  tags: [CreateTagInput!]!
}

input UpdateLocationInput {
  id: ID!
  name: String
  tags: [UpdateTagInput!]
}
