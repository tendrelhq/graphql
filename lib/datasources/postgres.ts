import type { Request } from "express";
import z from "myzod";
import postgres, { type Fragment } from "postgres";
import { makeActiveLoader } from "./activatable";
import makeAttachmentLoader from "./attachment";
import { makeAuditableLoader } from "./auditable";
import makeCustomerRequestedLanguageLoader from "./crl";
import { makeDescriptionLoader } from "./description";
import { makeDynamicStringLoader } from "./dynamicString";
import makeInvitationLoader from "./invitation";
import makeLanguageLoader from "./language";
import makeLocationLoader from "./location";
import {
  makeDisplayNameLoader,
  makeNameLoader,
  makeNameMetadataLoader,
} from "./name";
import makeOrganizationLoader from "./organization";
import { makeRequirementLoader } from "./requirement";
import { makeSopLoader } from "./sop";
import { makeStatusLoader } from "./status";
import makeTagLoader from "./tag";
import makeUserLoader from "./user";
import makeWorkerLoader from "./worker";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT: z.number({ coerce: true }).default(10),
  })
  .parse(process.env, { allowUnknown: true });

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  // TODO: this is probably a good idea...
  // connection: {
  //   statement_timeout: DB_STATEMENT_TIMEOUT * 1000, // milliseconds
  // },
});

export type SQL = typeof sql;
export type TxSQL = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}

export function orm(req: Request) {
  return {
    active: makeActiveLoader(req),
    attachment: makeAttachmentLoader(req),
    auditable: makeAuditableLoader(req),
    crl: makeCustomerRequestedLanguageLoader(req),
    description: makeDescriptionLoader(req),
    displayName: makeDisplayNameLoader(req),
    dynamicString: makeDynamicStringLoader(req),
    invitation: makeInvitationLoader(req),
    language: makeLanguageLoader(req),
    location: makeLocationLoader(req),
    name: makeNameLoader(req),
    nameMetadata: makeNameMetadataLoader(req),
    organization: makeOrganizationLoader(req),
    requirement: makeRequirementLoader(req),
    sop: makeSopLoader(req),
    status: makeStatusLoader(req),
    tag: makeTagLoader(req),
    user: makeUserLoader(req),
    worker: makeWorkerLoader(req),
  };
}

export type ORM = ReturnType<typeof orm>;
