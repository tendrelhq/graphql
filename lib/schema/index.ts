export * from "./__generated__/resolvers.generated";
export * from "./__generated__/typeDefs.generated";

export type Context = {
  authScope?: string;
  languageTypeId: number;
};
