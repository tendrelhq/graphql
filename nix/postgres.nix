{inputs, ...}: {
  imports = [
    inputs.process-compose.flakeModule
  ];
  perSystem = {
    config,
    self',
    pkgs,
    lib,
    ...
  }: {
    devShells.postgres = pkgs.mkShellNoCC {
      name = "tendrelhq/graphql+postgres";
      buildInputs = [
        config.packages.postgresql
        pkgs.perlPackages.TAPParserSourceHandlerpgTAP # pg_prove
        pkgs.python3.pkgs.sqlfmt
        pkgs.sqitchPg
        pkgs.squawk
      ];
    };

    packages = {
      inherit (pkgs) postgresql_14; # beta is on 14.12
      postgresql = pkgs.postgresql_16;
      pgddl = pkgs.stdenv.mkDerivation rec {
        name = "pgddl";
        version = "0.29";
        src = pkgs.fetchFromGitHub {
          owner = "lacanoid";
          repo = name;
          rev = version;
          hash = "sha256-W3G6TGtkj+zXXdGZZR0bmZhsLuFJvuGTlDoo8kL8sf0=";
        };
        buildInputs = with pkgs; [perl config.packages.postgresql];
        postPatch = ''
          patchShebangs .
        '';
        installPhase = ''
          install -D -t $out/share/postgresql/extension *.sql
          install -D -t $out/share/postgresql/extension *.control
        '';
      };
    };

    process-compose.devenv = {
      imports = [
        inputs.services.processComposeModules.default
      ];

      cli.options.no-server = false;

      services.postgres.pg1 = {
        enable = true;
        package = config.packages.postgresql;
        #
        extensions = exts:
          with exts;
          with config.packages; [
            pg_cron
            pgddl
            pgtap
            plpgsql_check
          ];
        hbaConf = lib.mkForce [
          # This one is the default docker bridge network.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.17.0.0/16";
            method = "trust";
          }
          # On my machine this is the tendrel_default network. YMMV.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.18.0.0/16";
            method = "trust";
          }
        ];
        initialDatabases = [
          {
            name = "dev";
          }
        ];
        settings = {
          listen_addresses = "0.0.0.0"; # so docker0 can connect
          log_statement = "all";
          logging_collector = false;
          shared_preload_libraries = "pg_cron,plpgsql_check";
        };
      };

      services.pgadmin.pga1 = {
        enable = true;
        extraConfig = {
          # Run in desktop mode with the master password disabled.
          MASTER_PASSWORD_REQUIRED = false;
          SERVER_MODE = false;
        };
        initialEmail = "";
        initialPassword = "";
      };

      settings.processes = {
        pga1 = {
          depends_on.pg1.condition = "process_healthy";
        };
      };
    };
  };
}
