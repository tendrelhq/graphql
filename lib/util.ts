import type { SortOrder } from "@/schema";
import { type GlobalId, decodeGlobalId } from "@/schema/system";
import { GraphQLError } from "graphql";
import type { Fragment } from "postgres";
import { sql } from "./datasources/postgres";

export function isError<T>(e: T | Error): e is Error {
  return e instanceof Error;
}

export function isValue<T>(v: T | Error): v is T {
  return !isError(v);
}

export function nullish<T>(t: T | null | undefined): t is null | undefined {
  return typeof t === "undefined" || t === null;
}

export class AssertionError extends Error {}

/**
 * Raises an assertion error when the condition does not hold, except in
 * production (and when DISABLE_ASSERTIONS is set) in which case we just log a
 * message instead of throwing.
 */
export function assert(condition: boolean, message = "assertion failed") {
  if (!condition) {
    if (
      // Only fire assertions when in dev/test environments,
      // and only when DISABLE_ASSERTIONS is unset.
      process.env.NODE_ENV !== "production" &&
      typeof process.env.DISABLE_ASSERTIONS === "undefined"
    ) {
      throw new AssertionError(message);
    }

    // Else just log a warning. We don't want to wholly lose our assertions.
    console.debug(`invariant violated: ${message}`);
  }
}

export function assertNonNull<T>(
  value: T | null | undefined,
  message = "Cannot return null for semantically non-nullable field.",
): T {
  if (nullish(value)) {
    throw new Error(message);
  }
  return value;
}

export function assertUnderlyingType(
  expected: string | string[],
  received: string,
) {
  const valid =
    typeof expected === "string"
      ? expected === received
      : expected.some(e => e === received);
  if (!valid) {
    throw new Error(
      `Invalid typename; expected: ${expected}, received: ${received}`,
    );
  }
  return received;
}

export type WithKey<T> = T & { _key: string };

export type PaginationArgs = {
  cursor?: GlobalId | null;
  direction: "forward" | "backward";
  limit: number;
};

type RawPaginationArgs = {
  // forward
  first?: number | null;
  after?: string | null;
  // backward
  last?: number | null;
  before?: string | null;
};

export function buildPaginationArgs(
  args: RawPaginationArgs,
  opts: {
    defaultLimit: number;
    maxLimit: number;
  },
): PaginationArgs {
  validatePaginationArgs(args);

  if (args.first) {
    return {
      cursor: map(args.after, decodeGlobalId),
      direction: "forward",
      limit: Math.min(args.first, opts.maxLimit),
    };
  }

  if (args.last) {
    return {
      cursor: map(args.before, decodeGlobalId),
      direction: "backward",
      limit: Math.min(args.last, opts.maxLimit),
    };
  }

  return {
    direction: "forward",
    limit: opts.defaultLimit,
  };
}

export function sortOrder(s: SortOrder): Fragment {
  switch (s) {
    case "asc":
      return sql`ASC`;
    case "desc":
      return sql`DESC`;
  }
}

export function validatePaginationArgs(args: RawPaginationArgs) {
  if ((args.after && args.before) || (args.first && args.last)) {
    throw new GraphQLError(
      "Invalid pagination arguments. To paginate forward, use the `first` and `after` arguments. To paginate backward, use the `last` and `before` arguments.",
      {
        extensions: {
          code: "BAD_REQUEST",
        },
      },
    );
  }
}

export function map<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
): T extends null | undefined ? T : R {
  if (nullish(t)) {
    return t as T extends null | undefined ? T : R;
  }
  return fn(t as NonNullable<T>) as T extends null | undefined ? T : R;
}

type OrElse<R> = () => R;
export function mapOrElse<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
  orElse: R | OrElse<R>,
): R {
  if (nullish(t)) {
    return typeof orElse === "function" ? (orElse as OrElse<R>)() : orElse;
  }
  return fn(t as NonNullable<T>);
}

export function inspect<T>(t: T) {
  console.debug("t =:", t);
  return t;
}

/**
 * Use this function to compare base64 encoded strings for identity.
 */
export function compareBase64(a: string, b: string): boolean {
  return normalizeBase64(a) === normalizeBase64(b);
}

// Postgres inserts newlines after 76 characters. Apparently it's a thing...
const newlinePattern = /\n/g;

/**
 * Use this function to normalize a base64 encoded string, i.e. remove newline
 * characters that are inserted by Postgres (for example) in certain cases.
 *
 * Note that most base64 implementations simply ignore these seemingly erroneous
 * newlines. As such, this function is mainly useful for (printf) debugging.
 */
export function normalizeBase64(s: string) {
  return s.replace(newlinePattern, "");
}
