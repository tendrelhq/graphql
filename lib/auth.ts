import { decodeGlobalId } from "@/schema/system";
import type { Context } from "@/schema/types";
import {
  ClerkExpressRequireAuth,
  type StrictAuthProp,
} from "@clerk/clerk-sdk-node";
import type e from "express";
import type { RequestHandler } from "express";
import { GraphQLError } from "graphql";
import * as jose from "jose";
import { type TxSql, sql } from "./datasources/postgres";
import { assert } from "./util";

declare global {
  namespace Express {
    interface Request extends StrictAuthProp {}
  }
}

export type Auth = StrictAuthProp["auth"];

export function clerk() {
  // In development it is nice to be able to use graphiql to help write your
  // queries. The default `devenv` will spin up a ruru[^1] graphiql endpoint
  // that you can use from your browser: http://localhost:1337.
  //
  // [^1]: https://github.com/graphile/crystal/tree/main/grafast/ruru
  //
  // tl;dr if you set the 'x-tendrel-user' header to a Clerk user id, we will
  // use that as the authenticated identity
  return (req: e.Request, res: e.Response, next: e.NextFunction) => {
    if (process.env.NODE_ENV === "development") {
      const userId = req.headers["x-tendrel-user"];
      console.log(`Backdoor hack engaged? ${userId}`);
      if (userId) {
        req.auth = {
          userId: userId as string,
        } as Auth;

        return next();
      }
    }

    ClerkExpressRequireAuth()(req, res, err => {
      if (err instanceof Error) {
        // @ts-ignore
        err.statusCode = 401;
      }

      next(err);
    });
  };
}

/**
 * Verify the provided *external* security token, generate a short-lived,
 * finely-grained subject token, and then call our /token api to perform the
 * actual token exchange. The client should not be aware that we are performing
 * this multi-step process.
 */
export const login: RequestHandler = async (req, res) => {
  // TODO: Move this elsewhere.
  const ALG = process.env.JWT_ALG ?? "HS256";
  const SECRET = new TextEncoder().encode(process.env.JWT_SECRET);
  // N.B. when JWT_ALG = HS256, JWT_SECRET must be at least 32 bytes!
  if (ALG === "HS256") assert(SECRET.length >= 32, "invalid jwt secret");

  const { userId } = req.auth;

  // TODO: ensure the user exists in the system, effectively making this API a
  // dual login/signup API. For now we will assume that the user has already
  // "signed up" for a Tendrel account, i.e. via the canonical Clerk/console flow.

  const temp = await new jose.SignJWT({ role: "anonymous" })
    .setProtectedHeader({ alg: ALG, typ: "JWT" })
    .setIssuer(`urn:tendrel:${process.env.STAGE}`)
    .setIssuedAt() // now
    .setNotBefore("-5 minutes")
    .setExpirationTime("5 minutes")
    .setSubject(userId)
    .sign(SECRET);

  try {
    // Same URL in both development and production. In development this will hit
    // the nginx proxy running on localhost. In production it will hit the same
    // proxy but running in a separate (ECS) container.
    const token = await fetch("http://localhost/api/v1/rpc/token", {
      method: "POST",
      body: JSON.stringify({
        grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
        subject_token: temp,
        subject_token_type: "urn:ietf:params:oauth:token-type:jwt",
      }),
      headers: {
        Authorization: `Bearer ${temp}`,
      },
    });

    const json = await token.json();

    res
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.1
      .setHeader("Cache-Control", "no-store")
      .setHeader("Pragma", "no-cache")
      .status(token.status);

    if (!token.ok) {
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.2
      res.json({
        error: json.code,
        error_description: json.message,
      });
    } else {
      res.json(json);
    }
  } catch (e) {
    console.error(e);
    res.status(500).send("Internal Server Error");
  }
};

export async function protect(
  { orgId, userId }: { orgId: string; userId: string },
  allowed: ("Worker" | "Supervisor" | "Admin")[],
) {
  const rows = await sql`
    select 1
    from public.workerinstance as w
    inner join public.systag as s on w.workerinstanceuserroleid = s.systagid
    where
        w.workerinstancecustomerid = (
            select customerid
            from public.customer
            where customeruuid = ${decodeGlobalId(orgId).id}
        )
        and w.workerinstanceworkerid = (
            select workerid
            from public.worker
            where workeridentityid = ${userId}
        )
        and s.systagtype in (${sql(allowed)});
  `;

  if (!rows.length) {
    throw new GraphQLError("Not authorized", {
      extensions: {
        code: "UNAUTHORIZED",
        hint: "You do not have the necessary permissions to perform this action",
      },
    });
  }
}

export function setCurrentIdentity(sql: TxSql, ctx: Context) {
  return sql`select * from auth.set_actor(${ctx.auth.userId}, ${ctx.req.i18n.language})`;
}
