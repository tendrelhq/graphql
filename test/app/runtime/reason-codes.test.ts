import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { setCurrentIdentity } from "@/auth";
import { type TxSql, sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import { decodeGlobalId } from "@/schema/system";
import type { Field } from "@/schema/system/component";
import { Task } from "@/schema/system/component/task";
import {
  createTestContext,
  execute,
  findAndEncode,
  setup,
} from "@/test/prelude";
import { assert, map } from "@/util";
import {
  CreateReasonCodeDocument,
  GetReasonCodeCompletionsDocument,
  ListReasonCodesDocument,
} from "./reason-codes.test.generated";

// Not yet ready for primetime :(
// `entity.import_entity` depends on some datawarehouse bullshit and I simply
// refuse to pull in that dependency here.
describe.skipIf(!!process.env.CI)("runtime + reason codes", () => {
  // See beforeAll for initialization of these variables.
  let CUSTOMER: string;
  let DOWN_TIME: Task;
  let IDLE_TIME: Task;

  // FIXME: This does not list associated templates, which is necessary in the
  // console view.
  test("list reason codes", async () => {
    // Name, Category (template)
    const result = await execute(schema, ListReasonCodesDocument, {
      // FIXME: Should not be required:
      owner: decodeGlobalId(CUSTOMER).id,
      // FIXME: This is not particularly ergonomic :/
      // Note that this is the entityinstanceuuid for the "Reason Code" systag:
      parent: ["d9b10b97-73aa-4407-948a-29f8434d525e"],
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test.todo("list reason codes (with filters)", async () => {
    // Filters: active
  });

  // FIXME: Ditto above: does not list associated templates.
  test.todo("create a new reason code", async () => {
    // Name, Category (template)
    const result = await execute(schema, CreateReasonCodeDocument, {
      name: "Overwhelming Confusion",
      // FIXME: Should not be required:
      owner: decodeGlobalId(CUSTOMER).id,
      parent: "d9b10b97-73aa-4407-948a-29f8434d525e",
      templates: [DOWN_TIME.id],
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test.todo("update a reason code", async () => {
    // e.g. rename, add to more templates?
  });

  test.todo("deactivate a reason code", async () => {
    // Not sure if necessary.
  });

  test.todo("delete a reason code", async () => {
    // Soft.
  });

  test("in the app, get completions", async () => {
    // For both Downtime and Idle Time.
    const r0 = await execute(schema, GetReasonCodeCompletionsDocument, {
      task: DOWN_TIME.id,
    });
    expect(r0.errors).toBeFalsy();
    expect(r0.data).toMatchSnapshot();

    const r1 = await execute(schema, GetReasonCodeCompletionsDocument, {
      task: IDLE_TIME.id,
    });
    expect(r1.errors).toBeFalsy();
    expect(r1.data).toMatchSnapshot();
  });

  beforeAll(async () => {
    // Setup:
    const ctx = await createTestContext();
    // 1. Create the demo customer
    const logs = await setup(ctx);
    CUSTOMER = findAndEncode("customer", "organization", logs);
    DOWN_TIME = map(
      findAndEncode("next", "worktemplate", logs),
      id => new Task({ id }),
    );
    IDLE_TIME = map(
      findAndEncode("next", "worktemplate", logs, { skip: 1 }),
      id => new Task({ id }),
    );
    await sql.begin(async sql => {
      await setCurrentIdentity(sql, ctx);

      // 2. Import the new customer into the entity model
      // FIXME: use Keller's API for customer create through the entity model.
      await sql`call entity.import_entity(null)`;

      const DOWN_TIME_CODES = [
        "Machine Down",
        "Waiting for Materials",
        "Scheduled Maintenance",
      ];
      const IDLE_TIME_CODES = ["Lunch Break", "Nothin to do!"];
      for (const [t, codes] of [
        [DOWN_TIME, DOWN_TIME_CODES],
        [IDLE_TIME, IDLE_TIME_CODES],
      ] as const) {
        // 3. Patch our template to have a 'Reason Code' field
        const field = await t.addField(ctx, {
          name: "Reason Code",
          type: "String",
        });

        // 4. Create demo reason codes for Downtime
        let order = 0;
        for (const code of codes) {
          await addReasonCodeToTemplate(
            sql,
            decodeGlobalId(CUSTOMER).id,
            code,
            order++,
            t,
            field,
          );
        }
      }
    });
  });

  afterAll(async () => {
    // Cleanup:
    // await cleanup(CUSTOMER);
    // 1. Delete reason codes
    // 2. Delete demo customer
    // 3. Call entity.import()?
  });
});

async function addReasonCodeToTemplate(
  sql: TxSql,
  customer: string,
  code: string,
  order: number,
  template: Task,
  field: Field,
) {
  assert(template._type === "worktemplate");
  const { id: fieldId, type: fieldType } = decodeGlobalId(field.id);
  assert(fieldType === "workresult");

  // Grab the owner id.
  const [{ owner }] = await sql`
    select entityinstanceuuid as owner
    from entity.entityinstance
    where entityinstanceoriginaluuid = ${customer}
  `;
  // Create the custag and entity_instance first.
  const [{ create_custaguuid: custag_id }] = await sql<
    [{ create_custaguuid: string }]
  >`
    call entity.crud_custag_create(
      ${owner},
      'd9b10b97-73aa-4407-948a-29f8434d525e',
      null,
      ${order},
      ${code},
      null,
      null,
      null,
      null, 
      null, 
      null,
      null,
      null,
      337::bigint
    );
  `;
  // Then create the template constraint.
  await sql`
    insert into public.worktemplateconstraint (
      worktemplateconstraintcustomerid,
      worktemplateconstraintcustomeruuid,
      worktemplateconstrainttemplateid,
      worktemplateconstraintresultid,
      worktemplateconstraintconstrainedtypeid,
      worktemplateconstraintconstraintid,
      worktemplateconstraintmodifiedby
    )
    select
      customerid,
      customeruuid,
      worktemplate.id,
      workresult.id,
      systag.systaguuid,
      custag.custaguuid,
      895
    from
      public.customer,
      public.worktemplate,
      public.workresult,
      public.systag,
      public.custag
    where customeruuid = ${customer}
      and worktemplate.id = ${template._id}
      and workresult.id = ${fieldId}
      and systag.systagtype = 'Reason Code'
      and custag.custaguuid = ${custag_id}
  `;
}
