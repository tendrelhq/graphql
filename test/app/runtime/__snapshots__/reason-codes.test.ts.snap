// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime + reason codes list reason codes 1`] = `
{
  "instances": {
    "__typename": "EntityInstanceConnection",
    "edges": [
      {
        "__typename": "EntityInstanceEdge",
        "asFieldTemplateValueType": {
          "__typename": "FieldConnection",
          "edges": [
            {
              "__typename": "FieldEdge",
              "node": {
                "__typename": "Field",
                "name": {
                  "__typename": "DisplayName",
                  "value": "Reason Code",
                },
                "parent": {
                  "__typename": "Task",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Downtime",
                  },
                },
              },
            },
          ],
        },
        "node": {
          "__typename": "EntityInstance",
          "name": {
            "__typename": "DisplayName",
            "value": "Machine Down",
          },
        },
      },
      {
        "__typename": "EntityInstanceEdge",
        "asFieldTemplateValueType": {
          "__typename": "FieldConnection",
          "edges": [
            {
              "__typename": "FieldEdge",
              "node": {
                "__typename": "Field",
                "name": {
                  "__typename": "DisplayName",
                  "value": "Reason Code",
                },
                "parent": {
                  "__typename": "Task",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Downtime",
                  },
                },
              },
            },
          ],
        },
        "node": {
          "__typename": "EntityInstance",
          "name": {
            "__typename": "DisplayName",
            "value": "Waiting for Materials",
          },
        },
      },
      {
        "__typename": "EntityInstanceEdge",
        "asFieldTemplateValueType": {
          "__typename": "FieldConnection",
          "edges": [
            {
              "__typename": "FieldEdge",
              "node": {
                "__typename": "Field",
                "name": {
                  "__typename": "DisplayName",
                  "value": "Reason Code",
                },
                "parent": {
                  "__typename": "Task",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Downtime",
                  },
                },
              },
            },
          ],
        },
        "node": {
          "__typename": "EntityInstance",
          "name": {
            "__typename": "DisplayName",
            "value": "Scheduled Maintenance",
          },
        },
      },
      {
        "__typename": "EntityInstanceEdge",
        "asFieldTemplateValueType": {
          "__typename": "FieldConnection",
          "edges": [
            {
              "__typename": "FieldEdge",
              "node": {
                "__typename": "Field",
                "name": {
                  "__typename": "DisplayName",
                  "value": "Reason Code",
                },
                "parent": {
                  "__typename": "Task",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Idle Time",
                  },
                },
              },
            },
          ],
        },
        "node": {
          "__typename": "EntityInstance",
          "name": {
            "__typename": "DisplayName",
            "value": "Lunch Break",
          },
        },
      },
      {
        "__typename": "EntityInstanceEdge",
        "asFieldTemplateValueType": {
          "__typename": "FieldConnection",
          "edges": [
            {
              "__typename": "FieldEdge",
              "node": {
                "__typename": "Field",
                "name": {
                  "__typename": "DisplayName",
                  "value": "Reason Code",
                },
                "parent": {
                  "__typename": "Task",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Idle Time",
                  },
                },
              },
            },
          ],
        },
        "node": {
          "__typename": "EntityInstance",
          "name": {
            "__typename": "DisplayName",
            "value": "Nothin to do!",
          },
        },
      },
    ],
  },
}
`;

exports[`runtime + reason codes in the app, get completions 1`] = `
{
  "node": {
    "__typename": "Task",
    "reasonCode": {
      "__typename": "Field",
      "completions": {
        "__typename": "ValueCompletionConnection",
        "edges": [
          {
            "__typename": "ValueCompletionEdge",
            "node": {
              "__typename": "ValueCompletion",
              "value": {
                "__typename": "StringValue",
                "string": "Waiting for Materials",
              },
            },
          },
          {
            "__typename": "ValueCompletionEdge",
            "node": {
              "__typename": "ValueCompletion",
              "value": {
                "__typename": "StringValue",
                "string": "Scheduled Maintenance",
              },
            },
          },
          {
            "__typename": "ValueCompletionEdge",
            "node": {
              "__typename": "ValueCompletion",
              "value": {
                "__typename": "StringValue",
                "string": "Machine Down",
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`runtime + reason codes in the app, get completions 2`] = `
{
  "node": {
    "__typename": "Task",
    "reasonCode": {
      "__typename": "Field",
      "completions": {
        "__typename": "ValueCompletionConnection",
        "edges": [
          {
            "__typename": "ValueCompletionEdge",
            "node": {
              "__typename": "ValueCompletion",
              "value": {
                "__typename": "StringValue",
                "string": "Lunch Break",
              },
            },
          },
          {
            "__typename": "ValueCompletionEdge",
            "node": {
              "__typename": "ValueCompletion",
              "value": {
                "__typename": "StringValue",
                "string": "Nothin to do!",
              },
            },
          },
        ],
      },
    },
  },
}
`;
